dmoj/judge.py

```diff
@@ -175,41 +179,202 @@ def begin_grading(
+            self.packet_manager.close()
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exception_type, exception_value, traceback):
+        pass
+
+
+class SubmissionJudgeProcess:
+    class BatchBegin:
+        pass
+
+    class BatchEnd:
+        pass
 
+    def __init__(self, problem_id, language, source, time_limit, memory_limit, short_circuit, meta):
+        self.problem_id = problem_id
+        self.language = language
+        self.source = source
+        self.time_limit = time_limit
+        self.memory_limit = memory_limit
+        self.short_circuit = short_circuit
+        self.meta = meta
+
+        self._terminate_grading = False
+        self.grader = None
+        self.process = None
+        self.process_conn = None
+
+    def run(self):
+        self.process_conn, child_conn = Pipe()
+        self.process = Process(target=self._run_grading_wrapper_process, args=(child_conn,))
+        self.process.start()
+
+        while True:
             try:
-                problem = Problem(problem_id, time_limit, memory_limit, meta)
-                self._block_and_grade(problem, language, source, short_circuit, report=report)
-            except Exception:
-                self.log_internal_error()
+                ipc_msg = self.process_conn.recv()
+            except OSError:
+                print("Failed to read message from grading subprocess", file=sys.stderr)
+                raise
 
-            self._terminate_grading = False
-            self.current_submission_id = None
-            self.current_submission_thread = None
-            self.current_grader = None
-            report(ansi_style('Done grading #ansi[%s](yellow)/#ansi[%s](green|bold).\n' % (problem_id, id)))
+            if ipc_msg is None:
+                break
 
-        self.current_submission_thread = threading.Thread(target=grading_cleanup_wrapper)
-        self.current_submission_thread.daemon = True
-        self.current_submission_thread.start()
+            yield ipc_msg
 
-        # Submission may have been killed already, but block if it hasn't been.
-        if blocking and self.current_submission_thread is not None:
-            self.current_submission_thread.join()
+        self.process_conn.close()
+        self.process.join()
+
+    def join(self):
+        self.process.join()
 
-    def grade_cases(self, grader, cases, short_circuit=False, is_short_circuiting=False):
+    def terminate_grading(self):
+        if self.process:
+            # TODO(tbrindus): can this race and fail?
+            self.process_conn.send(None)
+            self.join()
+
+    # TODO(tbrindus): do we need some sort of barrier here? what happens if an abort is sent before
+    # _block_and_grade is called?
+    def _run_abort_request_handler_thread(self, conn):
+        # If this ever returns, it's the signal that we should terminate the process.
+        # TODO(tbrindus): if it fails... it means our parent died?
+        conn.recv()
```

Comment:
When grading finishes, it doesn't look like we send anything before closing the `Pipe`. Will this print a pointless error?

Reply:
Actually this raced and never caused a problem because the process (and this thread) would exit before recv could raise. I've rejigged this all so there's a graceful teardown with two-sided ACK, even if the only message currently passed judge &rarr; sub-judge is abort.